Exclude test data from functions endpoint response

kshyju opened the pull request:
This PR modifies both read and write operations related to test data for functions.

1. When the `/admin/functions/{name}` PUT operation is called, the "TestData" property will no longer be saved. The response object for this call will not include "TestData" property populated.
2. When the `admin/functions/{name}` GET operation is called, the "TestData" property will no longer be included in the API response.

These changes are gated behind a feature flag, which is disabled (set to false) by default. We will begin by enabling this feature on a few stamps and, upon confirming its effectiveness, will roll it out more broadly. In a future release, we plan to remove the feature flag and enable this change by default.

### Pull request checklist

**IMPORTANT**: Currently, changes must be backported to the `in-proc` branch to be included in Core Tools and non-Flex deployments.

* [x] Backporting to the `in-proc` branch is not required
    * Otherwise: Link to backporting PR  - https://github.com/Azure/azure-functions-host/pull/10981
* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

kshyju committed:
Exclude test data from functions endpoint response.

kshyju commented:
Going to wait until https://github.com/Azure/azure-functions-host/issues/10954 is done.

kshyju committed:
Rename feature flag to be more generic (for READ and WRITE). Update write code path to check feature flag.

kshyju committed:
reset FileUtility.Instance after test

kshyju committed:
fix Property_ValidateAccess test

kshyju committed:
Merge branch 'dev' into shkr/remove-test_data

kshyju committed:
Adding a test for the write code path of the test data file.

kshyju committed:
Merge branch 'shkr/remove-test_data' of https://github.com/Azure/azure-functions-host into shkr/remove-test_data

kshyju committed:
Merge branch 'dev' into shkr/remove-test_data

fabiocav commented on src/WebJobs.Script.WebHost/Extensions/FunctionMetadataExtensions.cs:
```diff
@@ -54,7 +55,7 @@ public static async Task<FunctionMetadataResponse> ToFunctionMetadataResponse(th
             }
 
-            if (!string.IsNullOrEmpty(hostOptions.TestDataPath))
+            if (!string.IsNullOrEmpty(hostOptions.TestDataPath) && !excludeTestData)
```
Check the flag before checking the string here

kshyju replied:
Order swapped.

fabiocav commented on src/WebJobs.Script.WebHost/Management/WebFunctionsManager.cs:
```diff
@@ -33,9 +31,10 @@ public class WebFunctionsManager : IWebFunctionsManager
         private readonly IFunctionMetadataManager _functionMetadataManager;
         private readonly IHostFunctionMetadataProvider _hostFunctionMetadataProvider;
         private readonly IOptionsMonitor<LanguageWorkerOptions> _languageWorkerOptions;
+        private readonly IOptions<FunctionsHostingConfigOptions> _hostingConfigOptions;
```
Reviewing on mobile, but can you please check the scope of this service (web/jobhost) and consider using IOptionsMonitor here instead? Hosting config will be dynamically updated and the options refreshed when changed, but using IOptions will prevent you from getting updated states

kshyju replied:
Switched to `IOptionsMonitor<FunctionsHostingConfigOptions`.

We have 16 other instances using `IOptions<FunctionsHostingConfigOptions>`
https://github.com/search?q=repo%3AAzure%2Fazure-functions-host+IOptions%3CFunctionsHostingConfigOptions%3E&type=code&p=2

fabiocav commented on src/WebJobs.Script/Config/FunctionsHostingConfigOptions.cs:
```diff
@@ -147,6 +147,22 @@ internal bool EnableOrderedInvocationMessages
+        /// <summary>
+        /// Gets or sets a value indicating whether to ignore the TestData property during read and write operations of functions metadata.
+        /// </summary>
+        internal bool EnableTestDataSuppression
```
I realize this follows the convention of other properties here, but we should be using ‘Is[functionality]Enabled’

kshyju replied:
I switched to `IsTestDataSuppressionEnabled`

kshyju committed:
PR feedbak updates.
- Switched from `IOptions<FunctionsHostingConfigOptions>` to `IOptionsMonitor<FunctionsHostingConfigOptions>`
- renamed feature flag.

fabiocav commented:
@microsoft-github-policy-service rerun

mathewc commented:
Looks good - thanks Shyju!
